### DISCUSSION TITLE
Stanford CS143: Compilers - CS自学指南

### DISCUSSION THREAD

**Discussion by giscus** [PKUFlyingPig/cs-self-learning#134] [Announcements]
# Stanford CS143: Compilers - CS自学指南

CS自学指南

https://csdiy.wiki/%E7%BC%96%E8%AF%91%E5%8E%9F%E7%90%86/CS143/

**Comment by clickme-zxy** [comment]
虽然编译原理很重要，但是我一直不理解，为什么需要学这门课

**Reply by liuliming2019** [reply]
了解编译原理思想更要了解当前主要编译器才会产生强大作用，这样你才能根据编译器的实现机制去做语言如C++上面的优化，就像一些关键字、智能指针等实质上载编译器层面是做了怎样的处理，只有够了解才能更加优化。

**Reply by potatochick2020** [reply]
應該先學這還是 Operating system

**Reply by labyrinth-ssr** [reply]
感觉这两个没有依赖关系，我没学os先学的这个

**Reply by longguzzz** [reply]
学好了可以用AST生成代码。特定场景有需求的话，可以自己造一门专用语言。

**Reply by LiboShen** [reply]
我觉得其实如何写一个编译器这件事情本身并不重要。但重要的是，徒手写出一个编译器是一件非常非常难但却定义十分清晰资料非常充分的编程任务。所以通过学习最后完成大作业就是很有用的锻炼，而且也能收获巨大的成就感。

**Reply by wizardforcel** [reply]
> 虽然编译原理很重要，但是我一直不理解，为什么需要学这门课

等你需要自己折腾 DSL  做解析器的时候你就知道了。

**Reply by ShaneHeX** [reply]
> > 虽然编译原理很重要，但是我一直不理解，为什么需要学这门课
> 
> 等你需要自己折腾 DSL 做解析器的时候你就知道了。

还有就是性能分析调优吧

**Reply by dengyunsheng250** [reply]
JVM调优不也要学吗

**Reply by asiloh** [reply]
现在的 AI 架构都有用上编译技术进行中间优化，例如 Tensorflow、TVM，这些东西的本质就是一个内嵌的领域专用程序语言（EDSL）。 还有一个编程范式叫增量计算 （incremental computation），是反应式编程、数据流编程的一种，当节点上有任何数据更新/发送信号，数据相依的路径也会更新，游戏脚本设计、金融系统会用到。 还有树的优化，用元编程把树的走访消融在一块，可以减少快取丢失，浏览器的网页渲染会用得上。光追也有用元编程优化的技术。 所以编译原理学到的东西，不一定真的是要去搞编程语言设计还是编译器开发才用得上，做一些架构设计的时候它的精神本质就是一种编译器

**Reply by asiloh** [reply]
还有编译器工具链的build系统可以对应到很多的设计应用，可以看知乎这篇 https://zhuanlan.zhihu.com/p/375651053

**Reply by Juminiy** [reply]
数据库：ast，SQL优化器；
AI编译器；
服务端调优，对runtime，VM，自动GC等的调优。

**Reply by Juminiy** [reply]
前端工具链，造轮子。
汇编语言的设计哲学。

**Comment by Josh-Jing** [comment]
弱弱吐槽一句，这个bilibili链接是那个up主2020年上传的视频，字幕翻译几乎纯机翻。
建议英语还过得去的同学直接上2017年上传的这一版，内容大差不差，但是不容易受字幕影响走神：
https://www.bilibili.com/video/BV1ax411U78V?spm_id_from=333.337.search-card.all.click&vd_source=2d51412f9855fb2bb799971f101e0c46

**Reply by ShaneHeX** [reply]
edx 上面的不是更好？

**Reply by Josh-Jing** [reply]
wok，我不知道这玩意儿。搜了一下确实是个nb网站：https://www.edx.org/

**Reply by Delerrr** [reply]
我目前知道的三个课程资源网站：
- [MIT OpenCourseWare](https://ocw.mit.edu/)
- [Coursera](https://www.coursera.org/)
- [edX](https://www.edx.org/)

**Reply by pandacase** [reply]
确实，不过其实只要把翻译关了，只留一个英文字幕。跟其他那些在油管上面的课的体验是差不多的

**Reply by goahead656** [reply]
老哥，有视频链接嘛，我没找到视频在哪

**Comment by liubq7** [comment]
请问这个有什么前置要求吗

**Reply by Lockeded** [reply]
学一下形式语言与自动机比较好吧

**Comment by hondamio** [comment]
编译原理这门课我推荐国防科技大学这门 https://www.bilibili.com/video/BV12741147J3?p=1 讲的还可以不懂的地方基本上多看一两遍就懂了。

**Reply by goahead656** [reply]
老哥，视频挂了捏

**Comment by wuc9521** [comment]
强烈安利南京大学蚂蚁老师的编译原理, b站搜索ant-hengxin就可以搜到. 理论部分同样按照龙书的顺序覆盖了词法分析、语法分析、语义分析、运行时环境、寄存器分配、代码优化与生成等内容, 实践部分包括10~12个Lab, 而且每年课程进行一次大改.

**Reply by hondamio** [reply]
谢谢推荐！

**Reply by eightyun** [reply]
确实很好 可惜lab用的是java 我直接cpp泪目（悲）

**Reply by BoL0150** [reply]
这个lab公开吗？测试全吗？

**Reply by wuc9521** [reply]
@BoL0150 这个Lab的内容目前似乎没有对校外公开, 需要校园网访问, 不妨给老师发邮件

**Reply by zhoujunlingla** [reply]
请问大家跟着ant老师的课程，做cs143的lab这样可以吗

**Reply by Chasingwind-Z** [reply]
同问

**Reply by FineSixThuDB** [reply]
同问

**Reply by Dusker233** [reply]
可以给魏老师发邮件，我去年给魏老师发邮件要到了一个 oj 的账号和密码，但是当时没来得及做已经过期了（悲

**Reply by We-unite** [reply]
同问，有无老师的邮箱啥的

**Reply by Dusker233** [reply]
这学期蚂蚁老师的课向非 nju 同学开放了！具体可见课程首页 http://docs.compilers.cpl.icu/#/

**Reply by Kidding0204** [reply]
谢谢兄弟告知，oj账号很顺利的申请好了，处理速度很快，非南京大学的学生也能申请到。
蚂蚁老师，我敬爱你呀。

**Comment by ShaneHeX** [comment]
课程用 Ubuntu + VM 进行实验，在 VM 中进行大量的代码开发实在痛苦。
我做了一个 [Docker+vscode](https://github.com/hexinatgithub/stanford-compilers/tree/main/.devcontainer) 的开发环境，希望能帮助到大家

**Reply by EigenTrace** [reply]
404了

**Reply by ShaneHeX** [reply]
修复了

**Reply by ZefengBest** [reply]
"An error occured setting up the container" 有人遇到这个问题嘛

**Reply by ZefengBest** [reply]
大佬在build docker时出问题了
[+] Building 0.9s (6/8)                                    docker:desktop-linux
 => [internal] load build definition from Dockerfile-with-features         0.0s
 => => transferring dockerfile: 803B                                       0.0s
 => [internal] load .dockerignore                                          0.0s
 => => transferring context: 2B                                            0.0s
 => [internal] load metadata for mcr.microsoft.com/vscode/devcontainers/b  0.0s
 => [dev_container_auto_added_stage_label 1/5] FROM mcr.microsoft.com/vsc  0.0s
 => CACHED [dev_container_auto_added_stage_label 2/5] RUN apt-get update   0.0s
 => ERROR [dev_container_auto_added_stage_label 3/5] RUN apt-get install   0.9s
------

**Comment by wuyafei** [comment]
贡献一个用chatGPT翻译的完美中英双语版本：https://www.bilibili.com/video/BV1sV411g7Vv/?vd_source=2b9f6fcc0b53b5f5e1607a9e3e8c83bd

**Comment by eatvector** [comment]
新建了一个cs143qq群，大家感兴趣可以进群讨论学习：904654497

**Comment by dghtucs** [comment]
建了一个交流群，感兴趣可以进875607024

**Comment by ShaneHeX** [comment]
刚刚写完了所有的课程作业，给大家参考参考 [repo](https://github.com/hexinatgithub/stanford-compilers)

**Reply by codingMMax** [reply]
好兄弟 我点进去链接看 这门课的作业 需要stanford的学生账号 ？ 你们是怎么拿到作业的🤔️

**Comment by LuCC-que** [comment]
这节课的教材龙书好像已经有点过时了，cmu的那个编译课有人看过吗

**Comment by Coodist** [comment]
这玩意作业环境我自己配置的时候出现了一些障碍，如果你也有问题且无法解决可以加我qq211699660

**Comment by CodingKirby8620** [comment]
这门课在edX更新了新的压缩包，我还没做，暂时不知道有什么区别，有需要可以自取用：
2023/12/31版：
https://courses.edx.org/asset-v1:StanfordOnline+SOE.YCSCS1+1T2020+type@asset+block@student-dist.tar.gz
可以参考其他教程中“Directly install on linux”一项安装这个包
祝新年快乐

**Reply by lemonbiscuits-ljw** [reply]
感谢分享，新年快乐！

**Reply by CodingKirby8620** [reply]
更新：
每个作业是有grading script的，目前下面这个仓库是有收录的，大家可以去使用：
https://github.com/lennonwoo/CS143/tree/master

**Reply by rinevard** [reply]
现在好像有更新的版本了 `wget https://courses.edx.org/asset-v1:StanfordOnline+SOE.YCSCS1+3T2024+type@asset+block@student-dist.tar.gz`

**Comment by BasaErmi** [comment]
BV1Mb42177J7
-------
课程视频可以看b站上的这个，这是我目前看到翻译质量最好的cs143，应该是用gpt翻译的，而且英文字幕也重新修正了一遍

**Reply by BasaErmi** [reply]
补充一下，前半部分可以看这个：BV1cE411f78c，这是19年一位up主人工翻译的，但是只翻译到语法解析的bottom-up部分；后半部分可以看上面那个gpt翻译的

**Comment by dghtucs** [comment]
QQ群讨论873711918

**Comment by habaohaba** [comment]
视频里面的slides在哪里可以看到吗?
如果一个课程没有录像的话，怎么学习吗？纯看slides吗？
